Add/33 two column product page layout

louwie17 opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This adds several components to handle the product edit page. It is really simplistic as of now, and I think we should add slotfills within the `ProductFieldLayout` and `ProductFormLayout` at the end, so 3PD's can easily hook in with extra fields.

I am not set on the the implementation quite yet, and want to figure out a nicer way to pass fields in for each category. Also bear in mind that a category could have fields with an image on the side see the shipping dimensions section for example:
<img width="425" alt="Screen Shot 2022-07-28 at 4 42 41 PM" src="https://user-images.githubusercontent.com/2240960/181623868-9185f641-4311-43d4-a9a1-a4e155b3dee7.png">


Here is a quick demo of how it looks now:
<img width="1127" alt="Screen Shot 2022-07-28 at 4 33 17 PM" src="https://user-images.githubusercontent.com/2240960/181623927-3f7a3b4d-d9da-4e0d-ac98-5f5afcc04af0.png">

https://github.com/woocommerce/woocommerce/issues/34029

### How to test the changes in this Pull Request:

1. Enable the `new-product-management-experience` feature flag by using the WCA Test Helper (Tools > WCA Test Helper > Features)
2. Refresh the page and go to **Products > Add New (MVP)** and check if the layout matches that of the new design 5sAIeTRd9Yp7nSCT33BAWz-fi-5311%3A202025, fields don't need to match yet as those would be added in the future.
Also note that some of the `onChange` functions are empty callbacks, this is because the Form component isn't hooked up yet, this relies on https://github.com/woocommerce/woocommerce/pull/34082

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

louwie17 committed:
Add product page layout components

louwie17 committed:
Add a quick sample page to test the product form layout

louwie17 committed:
Add changelog

louwie17 commented on plugins/woocommerce-admin/client/products/edit-product-page.tsx:
```diff
@@ -0,0 +1,59 @@
+						description:
+							'This info will be displayed on the product page, category pages, social media, and search results.',
+						fields: [
+							<TextControl
```
I am not a 100% set on passing fields in like this, of-course these fields could be defined in a separate file and exported.

My other potential thought was to only pass in the field definition, like `name`, `label`, and `type` and render that within. I think this could be overkill though given a large amount of fields are unique it seems, the only overlapping one is the text field.

Instead of passing in a list of `fields` we could also just pass in one `CategorySection` component that is unique for each section. This doesn't feel very re-usable though (see description of main PR with example of shipping element where we don't only render fields but have an image on the right hand side).

Lastly we could even just define the category details here and remove the `fields` addition and replace that with a SlotFill within the `ProductFieldLayout` component (the only problem would be ordering of fields then, although we did solve this problem somewhat with the payment gateway slotfills)

mattsherman replied:
Some questions about extensibility that I don't have answered to. I think they may help shape how we define things though.

- Can an extension add a field in-between existing fields? I think that would be useful, though easily abused.

- Can an extension hide or override an existing field? Again, useful but easily abused.

mattsherman replied:
Feels kind of odd to pass the categories in through a prop instead of as child components. But, I guess passing as a prop would make it easier to define the categories externally and pull them in.

mattsherman replied:
Some example code for the different options would help to visualize what the dev experience would be.

louwie17 commented on plugins/woocommerce-admin/client/products/layout/product-form-layout.tsx:
```diff
@@ -0,0 +1,33 @@
+					/>
+					<ProductFieldLayout>{ fields }</ProductFieldLayout>
+				</div>
+			) ) }
```
A slotfill could be added after this `categories` map, or have a slotfill after each section, that would be named something to the effect of `{section_name}_after`

louwie17 commented on plugins/woocommerce-admin/client/products/layout/product-field-layout.tsx:
```diff
@@ -0,0 +1,8 @@
+import './product-field-layout.scss';
+
+export const ProductFieldLayout: React.FC = ( { children } ) => {
+	return <div className="product-field-layout">{ children }</div>;
```
A slotfill could be added here after rendering the children so 3PD's can render fields within a specific section.

mattsherman commented on plugins/woocommerce-admin/client/layout/controller.js:
```diff
@@ -24,6 +24,11 @@ import getReports from '../analytics/report/get-reports';
 
+const EditProductPage = lazy( () =>
+	import(
+		/* webpackChunkName: "analytics-report" */ '../products/edit-product-page'
```
This chunk name should not be this, correct?

octaedro committed:
Add option 2

louwie17 committed:
Refactor the product field layout structure

louwie17 committed:
Update page component structure of product form

louwie17 committed:
Add simple SlotFill support

louwie17 committed:
Update product page chunk name

mattsherman commented on plugins/woocommerce-admin/client/products/categories/product-details-category.tsx:
```diff
@@ -0,0 +1,31 @@
+				'woocommerce'
+			) }
+		>
+			<ProductFieldLayout name="Name">
+				<TextControl
+					label={ __( 'Name', 'woocommerce' ) }
+					value={ '' }
+					onChange={ () => {} }
+				/>
+			</ProductFieldLayout>
```
To avoid explicitly declaring the `ProductFieldLayout` wrapper around each control, we wrap the children of `ProductCategoryLayout` when rendering. See this quick and dirty CodePen for an example of how to wrap children and use props defined on them in the wrapper.

https://codepen.io/theminjam/pen/WNzMLxa?editors=0010

Also, why do we have to pass a noop function to `onChange`? I would think that should just be the default implementation of `onChange`.

Similar with `value`.

louwie17 replied:
I like the use of `Children` that sounds like a good plan. Although that will be based on the assumption that the right side of the category will be strictly fields, which may not always be the case (see shipping example in PR description for example).
Although I should be able to only wrap it if specific props exist.

Given this is just the layout I haven't hooked up the `onChange` callback yet, this actually depends on my form context provider PR - https://github.com/woocommerce/woocommerce/pull/34082

louwie17 replied:
Added the `Children` part in

mattsherman commented on plugins/woocommerce-admin/client/products/edit-product-page.tsx:
```diff
@@ -0,0 +1,57 @@
+import { ProductImagesCategory } from './categories/product-images-category';
+import { ProductFormLayout } from './layout/product-form-layout';
+
+const categories = [
```
This isn't used anymore.

mattsherman commented on plugins/woocommerce-admin/client/products/categories/product-details-category.tsx:
```diff
@@ -0,0 +1,31 @@
+				'woocommerce'
+			) }
+		>
+			<ProductFieldLayout name="Name">
```
Is the `name` prop unique globally, per category, or per something else? How is this enforced?

louwie17 replied:
Good point, I will add `categoryName` as an additional prop, as we probably won't have the same category name and field name.

mattsherman commented on plugins/woocommerce-admin/client/products/layout/product-field-layout.tsx:
```diff
@@ -0,0 +1,21 @@
+} ) => {
+	return (
+		<div className="product-field-layout">
+			<WooProductFieldItem.Slot fieldName={ name } location="before" />
```
How do multiple fills for a slot get ordered?

louwie17 replied:
First come first serve at the moment, I can add the order part back in that was also used in the `WooHeaderItem` example.

mattsherman reviewed (dismissed):
I like the direction this is going!

An incomplete review, but left some comments on the biggest questions and suggestions that jumped out!

louwie17 committed:
Add order to slot fill and move product field layout wrapper into product category layout

louwie17 committed:
Remove unused import

louwie17 commented:
@mattsherman if you have time today this should be ready for a re-review, the `on-change` is still blank as it depends on changes in this PR: https://github.com/woocommerce/woocommerce/pull/34082
And given this issue just deals with the layout, having the field hooked up is not necessary yet.

octaedro reviewed (dismissed):
Good job @louwie17! This is testing well here and the code looks good to me :shipit: 
Maybe as a follow-up PR we can add some tests to this.

joshuatf commented on plugins/woocommerce-admin/client/products/layout/woo-product-field-item.tsx:
```diff
@@ -0,0 +1,110 @@
+ * @param {Array} fills - slot's `Fill`s.
+ * @return {Node} Node.
+ */
+const sortFillsByOrder: Slot.Props[ 'children' ] = ( fills ) => {
```
Since we've also used this method in the `header` should we create a shared utils file?

louwie17 replied:
Yeah re-using it would probably make the most sense. All of these do have to be moved to a package at some point.

louwie17 replied:
Addressed here: https://github.com/woocommerce/woocommerce/pull/34113/commits/c2a2760fa2be3fad95c6a4cd8eefc2fa1b1fa95e

joshuatf commented on plugins/woocommerce-admin/client/products/categories/product-details-category.tsx:
```diff
@@ -0,0 +1,29 @@
+
+export const ProductDetailsCategory: React.FC = () => {
+	return (
+		<ProductCategoryLayout
```
Love this direction!

One nitpick I have is that `ProductCategory` made me think this was a "product category" (the taxonomy) at first.  Do these correlate with "sections" or some other semantic term to identify what this is?

louwie17 replied:
Good point, I like `Sections` that might be the better (less confusing) name to use.

louwie17 replied:
Addressed here: https://github.com/woocommerce/woocommerce/pull/34113/commits/a036816809d8e113282de8a20d79993f8d049e46

louwie17 committed:
Consolidate the slot fill ordering logic

louwie17 committed:
Rename category component to section component to prevent confusion

louwie17 committed:
Remove edit product page and use the new add product page instead

louwie17 commented:
@mattsherman @octaedro or @joshuatf could I get a re-review of any of you, I have updated this PR with the latest from @octaedro PR and addressed all the feedback.

louwie17 committed:
Add tests

louwie17 commented:
> @mattsherman @octaedro or @joshuatf could I get a re-review of any of you, I have updated this PR with the latest from @octaedro PR and addressed all the feedback.

I also end up adding some tests here: https://github.com/woocommerce/woocommerce/pull/34113/commits/1665cefa2fc4d4a1afc42e55276dd040c42ce190

octaedro reviewed (approved):
LGTM! Thank you for adding tests 🚀
